demos/portal-embed-demo/public/js/ttt/audio-controller.js

```diff
@@ -0,0 +1,369 @@
+    }
+
+    /**
+     * Showing the controller with anmimation
+     */
+    show() {
+        this.style.transition = 'bottom 0.2s';
+        this.style.bottom = '0px';
+    }
+
+    /**
+     * Adding anything required to do on portal activate
+     */
+    handleActivation() {
+        if(this._isPlaying()) {
+            // change the left transition position
+            const currentLeft = this._getAxisCurrentLeft();
+            const left = this.timeline.offsetWidth;
+            const duration = this.durationSec * (1 - currentLeft/left);
+            this.axis.style.transition = `${duration}s linear`;
+            this.axis.style.left = `${left}px`;
+        }
+    }
+
+    /**
+     * Adding event listeners for the buttons
+     */
+    _hookEvents() {
+        this.playButton.addEventListener('click', evt => {
+            if(this._isPlaying()){
+                this.pause();
+                return;
+            }
+            this.play();
+        });
+        this.nextButton.addEventListener('click', evt => this.next());
+        this.prevButton.addEventListener('click', evt => this.prev());
+        this.audio.addEventListener("ended", evt => {
+            this.pause();
+            this._createNewAudio(this.playlist[this.playlistIndex]);
+       });
+    }
+
+    /**
+     * Creating a new audio element
+     * @param {String} id - audio ID
+     */
+    _createNewAudio(id) {
+
+        // Simply replace the exisiting audio element with a new one
+        if(this.root.querySelector('audio')){
+            this.root.removeChild(this.root.querySelector('audio'));
+        }
+        const audio = document.createElement('audio');
+        const source = document.createElement('source');
+
+        // We could build streaming API for this in the future
+        source.src = `/mp3/${id}.mp3`;
+        audio.appendChild(source);
+        this.root.appendChild(audio);
+        this.audio = audio;
+
+        // Resetting the progress bar
+        this.durationSec = audioList[id].durationSec;
+        this.audioTitle.innerHTML
+            = `${audioList[id].title} - Totally Tooling Tips`;
+        this.axis.style.transition = '';
+        this.axis.style.left = '0px';
+    }
+
+    /**
+     * Update the prev/next button status
+     */
+    _updateController() {
+        if(this._hasPrev()) {
+            this.prevButton.style.opacity = 1.0;
+        } else {
+            this.prevButton.style.opacity = 0.3;
+        }
+        if(this._hasNext()){
+            this.nextButton.style.opacity = 1.0;
+        } else {
+            this.nextButton.style.opacity = 0.3;
+        }
+    }
+
+    /**
+     * Checks if the controller has a previous audio to play
+     * @returns {boolean}
+     */
+    _hasPrev() {
+        return this.playlistIndex !== 0;
+    }
+
+    /**
+     * Checks if the controller has a next audio to play
+     * @returns {boolean}
+     */
+    _hasNext() {
+        return this.playlist.length > 1 
```

Comment:
Is the first part of this condition actually necessary given the second part? ISTM that this should suffice:

```javascript
return this.playlistIndex < this.playlist.length - 1;
```

Reply:
True. Unnecessary. Fixed it to your suggestion.